Arm64: Implement VectorTableLookup/VectorTableLookupExtension intrinsinsic + Consecutive registers support

kunalspathak opened the pull request:
This adds support for remaining variant of `VectorTableLookup()` intrinsics that takes tuples for `table` parameter. The tuple can have 2, 3 or 4 `Vector128<byte>` values. These APIs has a requirement to allocate consecutive registers for the 1st operand value. This PR adds support for that as well.

### Details

During importer, depending on the form of 1st argument to the `VectorTableLookup()`, create a `FIELD_LIST` node with as many fields as number of entries in the `ValueTuple` of that argument. This node is further decomposed into local var.

In register allocation, during `buildInterval()`, we see this intrinsic and build as many `refPositions` as number of fields in the `ValueTuple`. We mark each of them with a flag `needsConsecutive` and for the first `refPosition`, we also set number of consecutive registers the entire series need (again, this is same as number of fields of `ValueTuple` in 1st argument). We also save the series in a newly added `nextConsecutiveRefPositionMap` map such that we could go to the next refposition in the series easily. (Note that `RefPosition` doesn't have a `next` pointer and adding that would have memory cost, hence I chose to add entries in a map). 

During register assignment, once the "first" `refPosition` gets a register, it sets the `registerAssignment` (Prior to the allocation pass, registerAssignment captures the valid registers for a `RefPosition`) of each subsequent `RefPosition`s of the series to the corresponding consecutive register(s) that comes after. Register assignment honors that decision (like it does today) and assigns those registers to the subsequent refpositions of that series.

### Example:

```c#
public static Vector128<byte> Test(float f)
    {
        var a = Produce1();
        var b = Produce2();
        var c = a + b;      
        var d = c + a;      
        var e = d + b;     
        d = AdvSimd.Arm64.VectorTableLookup((d, e, e, b), c); 
``` 

```asm
G_M53676_IG02:
            movz    x0, #0xD1FFAB1E      // code for helloworld:Produce1():System.Runtime.Intrinsics.Vector128`1[ubyte]
            movk    x0, #0xD1FFAB1E LSL #16
            movk    x0, #0xD1FFAB1E LSL #32
            ldr     x0, [x0]
            blr     x0
            str     q0, [fp, #0x20]	// [V01 loc0]
            movz    x0, #0xD1FFAB1E      // code for helloworld:Produce2():System.Runtime.Intrinsics.Vector128`1[ubyte]
            movk    x0, #0xD1FFAB1E LSL #16
            movk    x0, #0xD1FFAB1E LSL #32
            ldr     x0, [x0]
            blr     x0
            ldr     q16, [fp, #0x20]	// [V01 loc0]
            add     v17.16b, v16.16b, v0.16b
            str     q17, [fp, #0x10]	// [V03 loc2]
            add     v16.16b, v17.16b, v16.16b
            add     v18.16b, v16.16b, v0.16b
            mov     v17.16b, v18.16b
            mov     v19.16b, v0.16b
            ldr     q20, [fp, #0x10]	// [V03 loc2]
            tbl     v16.16b, {v16.16b, v17.16b, v18.16b, v19.16b}, v20.16b
            add     v0.16b, v0.16b, v16.16b
```

### TODO:
- [x] Explain the design in PR description 
- [x] Fix the summary docs for APIs
- [x] ~Get the `ValueTuple<>` APIs approved~ Do not expose the API through ref.
- [x] Add test coverage
- [x] Analyze and reduce throughput impact
- [x] Handle the round about case where `Rn` was assigned and the next refposition should get `R0`.

Contributes to https://github.com/dotnet/runtime/issues/1277, https://github.com/dotnet/runtime/issues/81599

kunalspathak committed:
Add VectorTableLookup 2/3/4 in hwinstrinsiclistarm64.h

kunalspathak committed:
Add VectorTableLookup

kunalspathak committed:
fixes to libraries

kunalspathak committed:
Prototype of simple tbl

kunalspathak committed:
Some progress

kunalspathak committed:
Some more updates

kunalspathak committed:
working model

kunalspathak committed:
Vector64<byte> support

kunalspathak committed:
Add VectorTableLookup_3

kunalspathak committed:
Add VectorTableLookup_4

kunalspathak committed:
cleanup

kunalspathak committed:
Remove regCount from LclVarDsc

kunalspathak committed:
Some more cleanup

Some more cleanup

kunalspathak committed:
setNextConsecutiveRegisterAssignment

kunalspathak committed:
Some more cleanup

kunalspathak committed:
TARGET_ARM64

kunalspathak committed:
Use getNextConsecutiveRefPositions instead of nextConsecutiveRefPosition field

kunalspathak committed:
jit format

kunalspathak committed:
Move getNextConsecutiveRefPosition

kunalspathak commented:
Here are few examples of C# and corresponding generated code: https://gist.github.com/kunalspathak/73922d385ea2642192c6167fa4a82778

SingleAccretion commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -1137,10 +1137,79 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                    {
+                        unsigned fieldCount = info.compCompHnd->getClassNumInstanceFields(sigReader.op1ClsHnd);
+                        op1->AsLclVar()->SetMultiRegUse();
+                        GenTreeFieldList* fieldList = new (this, GT_FIELD_LIST) GenTreeFieldList();
```
The field list is created... but not used?

kunalspathak replied:
True and I forgot to remove the code. I could get the struct's fields promoted by setting `lvUsedInSIMDIntrinsic = false`.

kunalspathak committed:
SA1141: Use tuple syntax

kunalspathak committed:
Remove the unwanted field list code

kunalspathak committed:
revert the flag that was mistakenly changed

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak commented:
@a74nh

kunalspathak committed:
Add test cases

kunalspathak committed:
FIELD_LIST

kunalspathak committed:
Use FIELD_LIST approach

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
jit format and fix arm build

kunalspathak committed:
fix assert failure

kunalspathak committed:
Add summary docs

Add summary docs in all the required files.

kunalspathak committed:
Make APIs public again

kunalspathak committed:
cleanup

kunalspathak committed:
Handle case for reg mod 32

kunalspathak committed:
Remove references from ref until API is approved

kunalspathak committed:
Use generic getFreeCandidates()

kunalspathak committed:
Add entries in ExtraAPis

kunalspathak committed:
Set CLSCompliant=false

tannergooding commented on src/libraries/System.Private.CoreLib/ref/System.Private.CoreLib.ExtraApis.cs:
```diff
@@ -37,3 +37,25 @@ public static partial class Debug
+        public static System.Runtime.Intrinsics.Vector64<sbyte> VectorTableLookup((System.Runtime.Intrinsics.Vector128<sbyte>, System.Runtime.Intrinsics.Vector128<sbyte>, System.Runtime.Intrinsics.Vector128<sbyte>, System.Runtime.Intrinsics.Vector128<sbyte>) table, System.Runtime.Intrinsics.Vector64<sbyte> byteIndexes) { throw null; }
+    }
+
+    public new abstract partial class Arm64 : System.Runtime.Intrinsics.Arm.ArmBase.Arm64
```
This should be nested under `AdvSimd`, not under the namespace.

kunalspathak committed:
Move in inner class

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
Remove CLSCompliant flag

tannergooding committed:
Add a suppression file for System.Runtime.Intrinsics on the new APIs until it they go through API review

kunalspathak commented:
Thanks @tannergooding  for the fix. It seemed to have worked.

The last thing remaining is understand and minimize the TP impact. This is happening because of adding checks for `needsConsecutive` in few methods of LSRA.

![image](https://user-images.githubusercontent.com/12488060/215036855-387e79dc-9712-417b-abcf-600ba8fa0894.png)

kunalspathak commented:
@dotnet/jit-contrib - This is ready for the review.

BruceForstall commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -1148,10 +1148,72 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                                intrinsic = (NamedIntrinsic)(intrinsic + 3);
+                                break;
+                            default:
+                                noway_assert("Unknown field count");
```
Missing `!`
```suggestion
                                noway_assert(!"Unknown field count");
```

BruceForstall commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -3035,7 +3035,7 @@ void LinearScan::UpdatePreferencesOfDyingLocal(Interval* interval)
 //    The node must not be contained, and must have been processed by buildRefPositionsForNode().
 //
-RefPosition* LinearScan::BuildUse(GenTree* operand, regMaskTP candidates, int multiRegIdx)
+RefPosition* LinearScan::BuildUse(GenTree* operand, regMaskTP candidates, int multiRegIdx, bool needsConsecutive)
```
Put `needsConsecutive` under `ARM64_ARG()`?

Add `needsConsecutive` to function header comment

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=149877&view=ms.vss-build-web.run-extensions-tab)

TP cost seems kind of high. But why is there any TP cost to non-ARM64?

BruceForstall reviewed (commented):
A few nits

SingleAccretion commented on src/coreclr/jit/codegenlinear.cpp:
```diff
@@ -1617,6 +1617,14 @@ void CodeGen::genConsumeRegs(GenTree* tree)
+            for (GenTreeFieldList::Use& use : tree->AsFieldList()->Uses())
+            {
+                GenTree* fieldNode = use.GetNode();
+                genConsumeReg(fieldNode);
```
```suggestion
                genConsumeRegs(fieldNode);
```
This function is not supposed to assume whether the child is node is contained or not.

kunalspathak replied:
done.

SingleAccretion commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -1148,10 +1148,72 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                    if (op1->TypeGet() == TYP_STRUCT)
+                    {
+                        assert(op1->OperIsLocal());
+
```
Can we actually assume `op1` is a local here? I would expect not, e. g. it could be an indir or a call. These cases would need to be handled by spilling a local (via `impAssignTempGen`).

Also, it would seem cleaner to me to put this intrinsic in the "requires special import" category and handle it in `hwintrinsicarm64.cpp`/`impSpecialIntrinsic` instead of here.

kunalspathak replied:
Done.

SingleAccretion commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -1148,10 +1148,72 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                            GenTreeLclFld* fldNode = gtNewLclFldNode(lclNum, TYP_SIMD16, offset);
+                            fieldList->AddField(this, fldNode, offset, TYP_SIMD16);
+
+                            offset += op1VarDsc->lvSize() / fieldCount;
```
```suggestion
                            offset += getTypeSize(TYP_SIMD16);
```
`lvSize` is not terribly cheap.

kunalspathak replied:
I wanted to generalize this because in future, we could also add APIs for `TYP_SIMD8`. What I can do is extract `p1VarDsc->lvSize() / fieldCount` out of the loop.

SingleAccretion replied:
If we want to generalize this, we would also want to change of the field that's created. So it could be: `offset += genTypeSize(fldNode);`.

kunalspathak replied:
I would know the `fldNode` type based on `fieldSize` which is `op1VarDsc->lvSize() / fieldCount`. I think I will just use `fieldSize` to forward the `offset`.

SingleAccretion commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -1148,10 +1148,72 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                            case 2:
+                                // NI_AdvSimd_VectorTableLookup_2
+                                // NI_AdvSimd_Arm64_VectorTableLookup_2
+                                intrinsic = (NamedIntrinsic)(intrinsic + 1);
```
Why do it this indirect way through addition instead of `intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_2`, etc?

kunalspathak replied:
because we want to set appropriate intrinsic based on if it is `AdvSimd` or `AdvSimd.Arm64`.

```c++
case 2:
  if (intrinsic == NI_AdvSimd_VectorTableLookup)
  {
     intrinsic = NI_AdvSimd_VectorTableLookup_2;
  }
  else
  {
     intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_2;
  }
  break;
case 3:
  if (intrinsic == NI_AdvSimd_VectorTableLookup)
  {
     intrinsic = NI_AdvSimd_VectorTableLookup_3;
  }
  else
  {
     intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_3;
  }
  break;
```

SingleAccretion commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -489,6 +504,40 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    // requirement, make sure to get the source's register because these
+                    // instruction encoding takes only the 1st register and infer the rest
+                    // from that.
+                    GenTree* op1 = firstField->AsCopyOrReload()->gtGetOp1();
+                    assert(!op1->IsCopyOrReload());
+                    op1Reg = op1->GetRegNum();
```
I am not sure I understand how this works. What guarantees the original register holds a valid value at this point (or does it not matter, or...)?

kunalspathak replied:
I am not sure if I follow. We create `RefPosition` based on the `GenTree` nodes, allocate registers to them and then write back the register assignments back to the tree nodes. If the register assigned at given point is different than what the original register it was assigned, we have a copy node, which we use to save the value in appropriate register at that point, before its use. This works just as any other place in codegen.

SingleAccretion replied:
What I do not understand here is why getting the register from the source of copy/reload is correct.

Let's consider the reload case. We have `firstField` assigned `REG_0` at def, it is spilled and the use has assigned `REG_1` to it. This should be:
```
t1 =  FIRST_FIELD_NODE REG_0
t2  = GT_RELOAD(t1) REG_1
      OUR_INTRINSIC(t2)
```
And here the code seemingly uses `REG_0` for `OUR_INTRINSIC`, what guarantess it will have the same value as `REG_1` that was loaded from the spill temp?

kunalspathak replied:
You are right, it doesn't. After my other changes, I don't think we need to handle the case for `CopyOrReload`. It should just use the node's register. I have also added an assert to make sure the subsequent refpositions gets consecutive registers.

kunalspathak commented:
> TP cost seems kind of high. But why is there any TP cost to non-ARM64?

It shouldn't be. I will investigate.

kunalspathak committed:
Review feedback

kunalspathak committed:
Add workaround for building tests

kunalspathak committed:
review feedback

kunalspathak committed:
TP: remove needsConsecutive parameter from BuildUse()

kunalspathak committed:
TP: Remove pseudo intrinsic entries

SingleAccretion commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1876,6 +1876,60 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+            argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg1, &argClass)));
+            op1     = impPopStack().val;
+
+            if (!op1->OperIsLocal())
```
```suggestion
            if (!op1->OperIs(GT_LCL_VAR))
```
`OperIsLocal` would return `true` for `LCL_FLD` as well, which the code doesn't handle.

SingleAccretion commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1876,6 +1876,60 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+
+            if (argType == TYP_STRUCT)
+            {
+                assert(op1->OperIsLocal());
+
+                LclVarDsc* op1VarDsc  = lvaGetDesc(op1->AsLclVar());
```
```suggestion
                LclVarDsc* op1VarDsc  = lvaGetDesc(op1->AsLclVar());
```
`AsLclVar` already contains a stricter assert.

SingleAccretion commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -489,6 +494,54 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+            case NI_AdvSimd_VectorTableLookup:
+            case NI_AdvSimd_Arm64_VectorTableLookup:
+            {
+
```
```suggestion
```

SingleAccretion commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -489,6 +494,54 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                {
+                    regCount++;
+#ifdef DEBUG
+
```
```suggestion
```

SingleAccretion commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -489,6 +494,54 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+
+                    GenTree* argNode = use.GetNode();
+                    assert(argReg == argNode->GetRegNum());
+                    argReg = (regNumber)(argReg + 1);
```
```suggestion
                    argReg = REG_NEXT(argReg);
```

SingleAccretion commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -2837,8 +2844,17 @@ regNumber LinearScan::allocateReg(Interval*    currentInterval,
     assignPhysReg(availablePhysRegRecord, currentInterval);
     refPosition->registerAssignment = foundRegBit;
+
+#ifdef TARGET_ARM64
```
Meta question on the `#ifdef`s: do we anticipate using this mechanism on other targets? Usually such things that are really target-independent but are not used on all targets are wrapped in another `#define`, e. g. `FEATURE_RARE_CONSECUTIVE_REGISTERS` (to give a hint it doesn't work that way for `TARGET_ARM/TYP_DOUBLE`...).

kunalspathak replied:
That's a fair point. I will add something.

SingleAccretion commented on src/coreclr/jit/lsra.h:
```diff
@@ -2259,6 +2279,14 @@ class RefPosition
     // would be 0..MAX_RET_REG_COUNT-1.
     unsigned char multiRegIdx : 2;
 
+#ifdef TARGET_ARM64
+    // If this refposition needs consecutive register assignment
+    bool needsConsecutive;
+
+    // How many consecutive registers does this and subsequent refPositions need
+    unsigned char regCount : 3;
+#endif // TARGET_ARM64
```
What is the impact on the size of `RefPosition` with these additional fields?

kunalspathak replied:
None. Before and after they remain 48 bytes.

  | Name | Value 
-- | -- | ---|
1.  | sizeof(Referenceable*) | 8 
2.  | sizeof(RefPosition*) | 8 
3.  | sizeof(GenTree*) | 8 
4.  | sizeof(bbNum) | 4 
5.  | sizeof(LsraLocation) | 4 
6.  | sizeof(regMaskTP) | 8 
7.  | sizeof(RefType) | 1 
8.  | sizeof(bit-fields) | 20 bits

SingleAccretion commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,65 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    return nextRefPosition;
+}
+
+//------------------------------------------------------------------------
+// setNextConsecutiveRegisterAssignment: Set the consecutive register mask to the
+//   subsequent refpositions
+//
+// Arguments:
+//    tree       - The GT_HWINTRINSIC node of interest
+//    pDstCount  - OUT parameter - the number of registers defined for the given node
+//
+// Return Value:
+//    The number of sources consumed by this node.
+//
+void LinearScan::setNextConsecutiveRegisterAssignment(RefPosition* firstRefPosition, regMaskTP firstRegAssigned)
```
Function header looks to not be in sync with what the function does.

SingleAccretion commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,65 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    assert(consecutiveRefPosition != nullptr);
+
+    regNumber firstReg    = genRegNumFromMask(firstRegAssigned);
+    regNumber regToAssign = firstReg == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(firstReg);
```
Does the function only handle FP registers?

(If so, an assert for that is missing)

SingleAccretion commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,65 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    regNumber regToAssign = firstReg == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(firstReg);
+
+#ifdef DEBUG
+    int refPosCount = 1;
```
There's an `INDEBUG` macro that can help trim down on `#ifdef DEBUG` code volume:
```
INDEBUG(int refPosCount = 1);
...
INDEBUG(refPosCount++);
```
If you like that kind of thing.

SingleAccretion commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -987,7 +1046,48 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+                            firstRefPos = currRefPos;
+                        }
+
+                        getNextConsecutiveRefPositionsMap()->Set(lastRefPos, currRefPos,
```
TP nit: cache the map in a local hoisted from the loop?

kunalspathak committed:
More fixes

kunalspathak commented:
> Thanks @tannergooding for the fix. It seemed to have worked.
> 
> The last thing remaining is understand and minimize the TP impact. This is happening because of adding checks for `needsConsecutive` in few methods of LSRA.
> 
> ![image](https://user-images.githubusercontent.com/12488060/215036855-387e79dc-9712-417b-abcf-600ba8fa0894.png)


TP regression is a bit lower. Most of the TP regression is coming because of adding `refPosition->needsConsecutive` checks in some of the hot methods like `allocateRegisters()` and `allocateReg()` method. Since this will only be triggered when `VectorTableLookup`  intrinsics API are used, ideally, the common path should not pay cost when compiling methods which doesn't have these APIs used (99.99% of the case). I will see if I can templatize some of this to eliminate the checks from methods that don't have these APIs.

![image](https://user-images.githubusercontent.com/12488060/216404541-b289396e-533f-4130-bb32-7a12686a1ac4.png)

kunalspathak committed:
Add the missing csproj

kunalspathak committed:
Fix test cases

kunalspathak committed:
Add fake lib for AdvSimd.Arm64* as well

kunalspathak committed:
Remove the workaround

kunalspathak committed:
Use template to control if consecutive registers is needed or not

kunalspathak committed:
jit format

kunalspathak committed:
fix the workaround

kunalspathak committed:
Revert "fix the workaround"

This reverts commit 1cb22d00d7159f40d33538ea29008e6764d52e45.

kunalspathak committed:
Revert "Remove the workaround"

This reverts commit b0b6a5e9abea195bd6dc35c5a638d280f20d6ed2.

kunalspathak commented:
> > Thanks @tannergooding for the fix. It seemed to have worked.
> > The last thing remaining is understand and minimize the TP impact. This is happening because of adding checks for `needsConsecutive` in few methods of LSRA.
> > ![image](https://user-images.githubusercontent.com/12488060/215036855-387e79dc-9712-417b-abcf-600ba8fa0894.png)
> 
> TP regression is a bit lower. Most of the TP regression is coming because of adding `refPosition->needsConsecutive` checks in some of the hot methods like `allocateRegisters()` and `allocateReg()` method. Since this will only be triggered when `VectorTableLookup` intrinsics API are used, ideally, the common path should not pay cost when compiling methods which doesn't have these APIs used (99.99% of the case). I will see if I can templatize some of this to eliminate the checks from methods that don't have these APIs.
> 
> ![image](https://user-images.githubusercontent.com/12488060/216404541-b289396e-533f-4130-bb32-7a12686a1ac4.png)

Latest results with templatizing the argument. I will investigate what arm64 TP is and why x64 is also showing regression.

![image](https://user-images.githubusercontent.com/12488060/216802908-b9d9eff8-454f-4e7c-965e-5a7439ad33d8.png)

kunalspathak committed:
Add VectorTableLookupExtensions  in libraries

kunalspathak committed:
Add support for VectorTableLookupExtension

kunalspathak committed:
WIP: available regs

kunalspathak committed:
WIP: Remove test hacks

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
Update getFreeCandidates() for consecutive registers

kunalspathak committed:
Add missing resetRegState()

kunalspathak committed:
Do not assume the current assigned register for consecutiveRegisters refposition is good.

If a refposition is marked as needConsecutive, then do not just assume that the existing register assigned
is good. We still go through the allocation for it to make sure that we allocate it a register such that the
consecutive registers are also free.

kunalspathak committed:
Handle case for copyReg

For copyReg, if we assigned a different register, do not forget to free the existing register it was holding

kunalspathak committed:
Update setNextConsecutiveRegister() with UPPER_VECTOR_RESTORE

kunalspathak committed:
Update code around copyReg

Updated code such that if the refPosition is already assigned a register, then
check if assignedRegister satisfies are needs (for first / non-first refposition).
If not, performs copyReg.

TODO: Extract the code surrounding and including copyReg until where we `continue`.

lambdageek committed:
Create the VectorTableLookup fake CoreLib as a reference assembly

Make the AdvSimd.Arm64 tests reference the VectorTableLookup fake
CoreLib as reference assembly; and ensure that it is not included as a
ProjectReference by the toplevel HardwareIntrinsics merged test
runners.

The upshot is that the AdvSimd.Arm64 tests can call the extra APIs via
a direct reference to CoreLib (instead of through System.Runtime), but
the fake library is not copied into any test artifact directories, and
the Mono AOT compiler never sees it.

That said, after applying this, the test fails during AOT compilation
of the *real* CoreLib

```
Mono Ahead of Time compiler - compiling assembly /Users/alklig/work/dotnet-runtime/runtime-bugs2/artifacts/tests/coreclr/osx.arm64.Release/Tests/Core_Root/System.Private.CoreLib.dll
  AOTID EA8D702E-9736-3BD5-435B-A9D5EEADCC78
  %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<byte>, System.Runtime.Intrinsics.Vector128`1<byte>>"* %arg_table
  <16 x i8>

  * Assertion: should not be reached at /Users/alklig/work/dotnet-runtime/runtime-bugs2/src/mono/mono/mini/mini-llvm.c:1455
```

lambdageek commented:
@kunalspathak Here's a partial fix for the mono AOT LLVM failure.

Basically it makes the VectorTableLookup.csproj produce a reference assembly only, and makes the AdvSimd.Arm64 tests only reference it as a reference assembly.

The upshot is that we ensure that the fake corelib never ends up in `artifacts/tests/` - it's only used during the managed build.  The AdvSimd.Arm64.dll ends up with a reference for System.Runtime.dll for all the normal public APIs and a direct reference to System.Private.CoreLib.dll for the "extra" APIs.

The mono AOT compilation then proceeds further.  However it then ends up hitting an assertion during AOT compilation of the real CoreLib.   I think we probably need Zoltan's help to figure out the right thing to avoid the assert.

1. Commit with the changes, feel free to pull into your PR and that other PR: https://github.com/lambdageek/runtime/commit/347d9437162858f2c48237547a3ed9b810ef2f5a

2. It might be convenient to make all these fake ref assemblies follow a common naming pattern for the .csproj file.  like DivRem.RefOnly.csproj.   Then HardwareIntrinsics_r.csproj can just do `<ProjectReference Include="**/*.csproj" Exclude="**/*.RefOnlly.csproj">` and the individual test assemblies can do the project references like in my commit.  That way if we need more fake CoreLibs, we can make it work without a lot of boilerplate.

3. The AOT failure in the real corelib is:
```
 Mono Ahead of Time compiler - compiling assembly /Users/alklig/work/dotnet-runtime/runtime-bugs2/artifacts/tests/coreclr/osx.arm64.Release/Tests/Core_Root/System.Private.CoreLib.dll
  AOTID EA8D702E-9736-3BD5-435B-A9D5EEADCC78
  %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<byte>, System.Runtime.Intrinsics.Vector128`1<byte>>"* %arg_table
  <16 x i8>

  * Assertion: should not be reached at /Users/alklig/work/dotnet-runtime/runtime-bugs2/src/mono/mono/mini/mini-llvm.c:1455
```

kunalspathak commented:
> The mono AOT compilation then proceeds further. However it then ends up hitting an assertion during AOT compilation of the real CoreLib. I think we probably need Zoltan's help to figure out the right thing to avoid the assert.

FYI - @vargaz 

Is there a tracking issue for this? I believe I will have to wait until it is fixed so I don't break the CI?

kunalspathak committed:
Rename VectorTableLookup to VectorTableLookup.RefOnly

vargaz commented:
This will fix the aot crash:
https://gist.github.com/vargaz/2aad0dabbb7ef2424fd4f2d3da3d4e5d

kunalspathak commented:
> This will fix the aot crash: https://gist.github.com/vargaz/2aad0dabbb7ef2424fd4f2d3da3d4e5d

The previous and the new code just handles 1 variant of VectorTableLookup/VectorTableLookupExtension i.e. TBL1/TBX1. What about other variants or it doesn't matter here?

```diff
diff --git a/src/mono/mono/mini/simd-intrinsics.c b/src/mono/mono/mini/simd-intrinsics.c
index 58752308d57..0721f20d7d2 100644
--- a/src/mono/mono/mini/simd-intrinsics.c
+++ b/src/mono/mono/mini/simd-intrinsics.c
@@ -437,6 +437,12 @@ is_hw_intrinsics_class (MonoClass *klass, const char *name, gboolean *is_64bit)
 	}
 }
 
+static gboolean
+type_is_simd_vector (MonoType *type)
+{
+	return type->type == MONO_TYPE_GENERICINST && m_class_is_simd_type (mono_class_from_mono_type_internal (type));
+}
+
 static MonoTypeEnum
 get_underlying_type (MonoType* type)
 {
@@ -2879,7 +2885,7 @@ static SimdIntrinsic advsimd_methods [] = {
 	{SN_TransposeOdd, OP_ARM64_TRN2},
 	{SN_UnzipEven, OP_ARM64_UZP1},
 	{SN_UnzipOdd, OP_ARM64_UZP2},
-	{SN_VectorTableLookup, OP_XOP_OVR_X_X_X, INTRINS_AARCH64_ADV_SIMD_TBL1},
+	{SN_VectorTableLookup},
 	{SN_VectorTableLookupExtension, OP_XOP_OVR_X_X_X_X, INTRINS_AARCH64_ADV_SIMD_TBX1},
 	{SN_Xor, OP_XBINOP_FORCEINT, XBINOP_FORCEINT_XOR},
 	{SN_ZeroExtendWideningLower, OP_ARM64_UXTL},
@@ -3184,6 +3190,10 @@ emit_arm64_intrinsics (
 			ret->sreg3 = scalar->dreg;
 			return ret;
 		}
+		case SN_VectorTableLookup:
+			if (!type_is_simd_vector (fsig->params [0]))
+				return NULL;
+			return emit_simd_ins_for_sig (cfg, klass, OP_XOP_OVR_X_X_X, INTRINS_AARCH64_ADV_SIMD_TBL1, 0, fsig, args);
 		default:
 			g_assert_not_reached ();
 		}
```

vargaz commented:
Would be better to handle them as well, but this change fixes the crash in the aot compiler for now.

kunalspathak committed:
Start consecutive refpositions with RefTypeUse and never with RefTypeUpperVectorSave

kunalspathak committed:
Add test cases for VectorTableLookupExtension

kunalspathak committed:
Pass the missing defaultValues

kunalspathak committed:
Use platform neutral BitScanForward

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
jit format

tannergooding commented:
@kunalspathak, is this "ready"? It looks like there are some formatting errors and CI failures for Mono.

There is also still a large TP regression for Arm64, particularly around `minopts`.

tannergooding commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -2125,6 +2129,8 @@ void LinearScan::buildIntervals()
     }
 #endif // DEBUG
 
+    resetRegState();
```
What's the reason for needing to `resetRegState()` when doing `buildIntervals` that we didn't need before?

tannergooding commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -987,11 +1228,41 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
     }
 
-    if ((intrin.category == HW_Category_SIMDByIndexedElement) && (genTypeSize(intrin.baseType) == 2))
+    if ((intrin.id == NI_AdvSimd_VectorTableLookup) || (intrin.id == NI_AdvSimd_Arm64_VectorTableLookup) ||
```
I expect `VectorTableLookup` to be rarer than the others, so it might be worth putting it after `SimdByIndexedElement`

It might also be worth a flag or category to avoid needing 4-6 separate checks each time

kunalspathak replied:
> It might also be worth a flag or category to avoid needing 4-6 separate checks each time

Not sure if I follow this. Can you elaborate?

tannergooding replied:
This used to be `if ((intrin.category == HW_Category_SIMDByIndexedElement) && (genTypeSize(intrin.baseType) == 2))` checked first which covers 51 different intrinsics.

Now, we're instead doing the checks against `VectorTableLookup` first and that's going to be "less common" overall.

This means the other 51 intrinsics now have to go through 4 additional branches before their path can be hit.

It would be better to handle `SIMDByIndexedElement` first and then handle `VectorTableLookup` as the `else if` block instead.

It would also be good to reduce the number of checks required to determine if this branch should be taken at all. If we mark these with a special category or flag, then we can do 1 check that covers all 4 intrinsics.

tannergooding commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,247 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//
+bool LinearScan::areNextConsecutiveRegistersFree(regNumber regToAssign, int registersCount, var_types registerType)
+{
+    for (int i = 0; i < registersCount; i++)
+    {
+        if (isRegInUse(regToAssign, registerType))
+        {
+            return false;
+        }
+        regToAssign = regToAssign == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(regToAssign);
+    }
```
We could presumably simplify this down to not loop. I believe the following are correct given the general algorithm for finding consecutively set bits:

To find the two consecutive bits:
```cpp
x &= (x >> 1);
```

To find three consecutive bits:
```cpp
x &= (x >> 1);
x &= (x >> 1);
```

To find four consecutive bits:
```cpp
x &= (x >> 2);
x &= (x >> 1);
```

For each of these, `x != 0` means there exists a sequence of 2/3/4 bits and the remaining set bit is the start of the set sequence.

Consider finding 3 consecutive bits in `9952`, which is `0b0010_0110_1110_0000`. This will end up producing `32`, which is `0b0010_0000` and hence we know that there is a sequence of bits and it is `0b1110_0000`

kunalspathak committed:
Remove the fake testlib workaround

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
Fix mono failures

kunalspathak committed:
Fix x64 TP regression

kunalspathak committed:
Fix test cases

kunalspathak committed:
fix some more tp regression

kunalspathak committed:
Fix test build

kunalspathak committed:
misc. changes

kunalspathak committed:
Fix the bug where we were not freeing copyReg causing an assert in tier0

kunalspathak committed:
Refactor little bit to reduce checks for VectorTableLookup

kunalspathak commented:
Latest minopts diffs locally:

```
Base: 2637303702, Diff: 2640095957, +0.1059%

??$allocateRegisters@$0A@@LinearScan@@QEAAXXZ                                                 : 67890538  : NA       : 45.06% : +2.5742%
?fgDfsInvPostOrderHelper@Compiler@@IEAAXPEAUBasicBlock@@AEAPEA_KPEAI@Z                        : 3565144   : NA       : 2.37%  : +0.1352%
?getFreeCandidates@LinearScan@@AEAA_K_KPEAVRefPosition@@@Z                                    : 1914990   : NA       : 1.27%  : +0.0726%
?fgDfsInvPostOrder@Compiler@@IEAAXXZ                                                          : 1567162   : NA       : 1.04%  : +0.0594%
?processBlockEndAllocation@LinearScan@@AEAAXPEAUBasicBlock@@@Z                                : 685123    : NA       : 0.45%  : +0.0260%
?newRefPositionRaw@LinearScan@@AEAAPEAVRefPosition@@IPEAUGenTree@@W4RefType@@@Z               : 657151    : +2.33%   : 0.44%  : +0.0249%
?allocateReg@LinearScan@@AEAA?AW4_regNumber_enum@@PEAVInterval@@PEAVRefPosition@@@Z           : 180593    : +0.17%   : 0.12%  : +0.0068%
?getRangeEndLocation@RefPosition@@QEAAIXZ                                                     : -153831   : -7.78%   : 0.10%  : -0.0058%
?newRefPosition@LinearScan@@AEAAPEAVRefPosition@@PEAVInterval@@IW4RefType@@PEAUGenTree@@_KI@Z : -243216   : -1.25%   : 0.16%  : -0.0092%
?fgDfsReversePostorder@Compiler@@IEAAXXZ                                                      : -1553354  : -100.00% : 1.03%  : -0.0589%
?fgDfsReversePostorderHelper@Compiler@@IEAAXPEAUBasicBlock@@AEAPEA_KAEAI2@Z                   : -3745444  : -100.00% : 2.49%  : -0.1420%
?allocateRegisters@LinearScan@@QEAAXXZ                                                        : -67970807 : -100.00% : 45.11% : -2.5773%
```

kunalspathak committed:
Add template parameter for allocateReg/copyReg/select

kunalspathak commented:
@vargaz - any idea on what could be causing these failures? 

```
[System.Private.CoreLib.dll] Exec (with response file contents expanded) in /tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish: MONO_PATH=/tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish: MONO_ENV_OPTIONS= /tmp/helix/working/C27F0AC8/p/build/cross/mono-aot-cross -O=gsharedvt --debug --llvm "--aot=mtriple=arm64-ios,static,dwarfdebug,mattr=+crc,direct-icalls,nimt-trampolines=2000,ntrampolines=40000,nrgctx-fetch-trampolines=256,ngsharedvt-trampolines=4400,nftnptr-arg-trampolines=4000,nrgctx-trampolines=40000,nodebug,llvm-path=/tmp/helix/working/C27F0AC8/p/build/cross,static,dedup-skip,full,asmonly,outfile=/tmp/helix/working/C27F0AC8/w/BC510A3E/e/apple_build/obj/mobile/System.Private.CoreLib.dll.s,llvm-outfile=/tmp/helix/working/C27F0AC8/w/BC510A3E/e/apple_build/obj/mobile/System.Private.CoreLib.dll-llvm.o,data-outfile=/tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish/System.Private.CoreLib.aotdata" "System.Private.CoreLib.dll"
  
/tmp/helix/working/C27F0AC8/p/build/apple/AppleApp.targets(110,5): error : Precompiling failed for /tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish/System.Private.CoreLib.dll with exit code 139. [/private/tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/C27F0AC8/p/build/apple/AppleApp.targets(110,5): error :  [/private/tmp/helix/working/C27F0AC8/w/BC510A3E/e/publish/ProxyProjectForAOTOnHelix.proj]
```

kunalspathak committed:
Comments

SamMonoRT commented:
/cc @fanyang-mono

fanyang-mono commented on src/mono/mono/mini/simd-intrinsics.c:
```diff
@@ -3365,6 +3371,14 @@ emit_arm64_intrinsics (
+        case SN_VectorTableLookup:
+			if (!type_is_simd_vector (fsig->params [0]))
+				return NULL;
+			return emit_simd_ins_for_sig (cfg, klass, OP_XOP_OVR_X_X_X, INTRINS_AARCH64_ADV_SIMD_TBL1, 0, fsig, args);
```
I am curious if this change is necessary. If I understand this change correctly, the only thing changed was the value of `inst_c1` field, which was actually ignored by `OP_XOP_OVR_X_X_X` in mini-llvm.c.

fanyang-mono replied:
Discussed offline.

vargaz commented:
This will fix the corlib failure:
https://gist.github.com/vargaz/a86b3f6bd265786a9c89981652dff9bc

kunalspathak committed:
Fix mono failures

kunalspathak commented on src/mono/mono/mini/simd-intrinsics.c:
```diff
@@ -434,6 +434,12 @@ emit_simd_ins_for_unary_op (MonoCompile *cfg, MonoClass *klass, MonoMethodSignat
+static gboolean
+type_is_simd_vector (MonoType *type)
+{
+	return type->type == MONO_TYPE_GENERICINST && m_class_is_simd_type (mono_class_from_mono_type_internal (type));
```
robably @vargaz can answer those questions.

fanyang-mono replied:
Discussed offline.

vargaz commented:
Having tuples as arguments could lead to some code bloat on AOT platforms, since the AOT compiler would have to generate code for the virtual methods on these types like ToString() etc. which can get quite large.

kunalspathak commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -5329,34 +5371,139 @@ void LinearScan::allocateRegisters()
+                        // one that was assigned previously.
+
+                        regsInUseThisLocation |= copyRegMask;
+                        if (currentRefPosition.lastUse)
```
I will refactor some of this code to move into a method, because it is used at two places.

kunalspathak committed:
Added some more comments

tannergooding commented:
> Having tuples as arguments could lead to some code bloat on AOT platforms, since the AOT compiler would have to generate code for the virtual methods on these types like ToString() etc. which can get quite large.

This is the way that hardware intrinsic APIs were designed and implemented to handle all this. Moving away from tuples would be very undesirable. We'd end up having the "same code" in any explicitly defined type (an explicit `ToString`, `Equals`, and `GetHashCode` implementation; implementing `IEquatable<T>`, etc...) and so this really just ends up being something that AOT and/or trimmer should handle when the types are "used".

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    regMaskTP overallResult          = RBM_NONE;
+    regMaskTP consecutiveResult      = RBM_NONE;
+    DWORD     regAvailableStartIndex = 0, regAvailableEndIndex = 0;
+    do
```
This algorithm for finding sequences of registers doesn't consider "wrapping", like your allocation code does -- so that allocation code wrapping will never get run (presumably).

kunalspathak replied:
Yes, that is a known problem and I think I have written a TODO somewhere. If not, will add a comment unless I can find an easy way to add it.

kunalspathak replied:
Fixed this now.

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+#define AppendConsecutiveMask(startIndex, endIndex, availableRegistersMask)                                            \
+    regMaskTP selectionStartMask = (1ULL << regAvailableStartIndex) - 1;                                               \
+    regMaskTP selectionEndMask   = (1ULL << (regAvailableEndIndex - registersNeeded + 1)) - 1;                         \
+    consecutiveResult |= availableRegistersMask & (selectionEndMask & ~selectionStartMask);                            \
```
Isn't it the case here that `availableRegistersMask` *must* have all the bits set that `(selectionEndMask & ~selectionStartMask)` has set?

kunalspathak replied:
That's right, and we want to just extract the 1st of those bits to be set in `consecutiveResult `

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+        return RBM_NONE;
+    }
+
+// At this point, for 'n' registers requirement, if Rm+1, Rm+2, Rm+3, ..., Rm+k are
```
Shouldn't this be `Rm, Rm+1, Rm+2, ..., Rm + k - 1`?

kunalspathak commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -2277,6 +2287,8 @@ void LinearScan::checkLastUses(BasicBlock* block)
                     foundDiff = true;
                 }
+
+                JITDUMP("++ V%02u in computedLive\n", compiler->lvaTrackedIndexToLclNum(varIndex));
```
I was debugging something and added it then. But I can remove it, don't think much useful.

kunalspathak commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -2831,11 +2844,24 @@ bool LinearScan::isMatchingConstant(RegRecord* physRegRecord, RefPosition* refPo
-    regMaskTP foundRegBit = regSelector->select(currentInterval, refPosition DEBUG_ARG(registerScore));
+    regMaskTP foundRegBit;
+
+#ifdef TARGET_ARM64
+    if (hasConsecutiveRegister)
+    {
+        foundRegBit = regSelector->select<true>(currentInterval, refPosition DEBUG_ARG(registerScore));
+    }
+    else
+#endif // TARGET_ARM64
+    {
+        foundRegBit = regSelector->select(currentInterval, refPosition DEBUG_ARG(registerScore));
+    }
```
> by always defining the template,

You mean for other platforms too? I didn't want to bloat the size of other platforms as well because of that.

BruceForstall replied:
Will it bloat the other platforms if only `false` is ever used for them?

kunalspathak replied:
i am not sure. I will try it out and confirm.

kunalspathak replied:
Before change:

<img width="551" alt="image" src="https://user-images.githubusercontent.com/12488060/227102773-dc3a3764-4396-4283-bd3f-6462f722f59c.png">

After change:

<img width="590" alt="image" src="https://user-images.githubusercontent.com/12488060/227102734-23e2f0fd-d127-4f2e-a8d8-87a532118984.png">

I will go ahead and use your suggestion.

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
+                consecutiveRefPosition->registerAssignment &= ~busyConsecutiveRegMask;
+            }
+            consecutiveRefPosition = getNextConsecutiveRefPosition(consecutiveRefPosition);
```
can you elaborate what you mean?

BruceForstall replied:
For `consecutiveRefPosition->refType == RefTypeUpperVectorRestore` you call `consecutiveRefPosition = getNextConsecutiveRefPosition(consecutiveRefPosition);` exactly once and then assume (assert) the next ref position is the RefTypeUse you're looking for. I guess that's guaranteed to occur in that order?

kunalspathak replied:
Yes, because `BuildConsecutiveRegistersForUse()` checks for that while storing it in the refPositionMap.

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
+//------------------------------------------------------------------------
+// getFreeCandidates: Returns the mask of all the free candidates for given refPosition.
+//   If refPosition is the first RefPosition of a series of refpositions that needs
+//   consecutive registers, then returns only the mask such that it satisfies the need
+//   of having free consecutive registers.
```
> does it only include the first register for any consecutive register sequence

Only the first register or first (n - k) registers, if there are `n` registers available and we need just `k` consecutive registers.

BruceForstall replied:
Maybe you can expand the comment? (Maybe even with an example?)

kunalspathak replied:
Done.

kunalspathak committed:
Call allocateReg/assignCopyReg/select methods only for refpositions that need consecutive registers

kunalspathak committed:
Add heuristics to pick best possible set of registers which will need less spilling

kunalspathak committed:
setNextConsecutiveRegisterAssignment() no longer checks for areNextConsecutiveRegistersFree()

kunalspathak committed:
Rename getFreeCandidates() -> getConsecutiveCandidates()

kunalspathak committed:
fix parameters to areNextConsecutiveRegistersFree()

kunalspathak committed:
Rename and update canAssignNextConsecutiveRegisters()

kunalspathak committed:
Add the missing setNextConsecutiveRegisterAssignment() calls

kunalspathak committed:
Fix a condition for upperVector

kunalspathak committed:
Update spill heurisitic to handle cases for jitstressregs

kunalspathak committed:
Misc. remove popcount() check from getConsecutiveRegisters()

kunalspathak committed:
jit format

kunalspathak committed:
Fix a bug in canAssignNextConsecutiveRegisters()

kunalspathak committed:
Add filterConsecutiveCandidates() and perform free/busy candidates scan

kunalspathak committed:
Consume the new free/busy consecutive candidates method

kunalspathak committed:
Handle case where 'copyReg == assignedReg'

kunalspathak committed:
Misc. cleanup

kunalspathak committed:
Include LsraExtraFPSetForConsecutive for stress regs

kunalspathak committed:
handle case where 'assignedInterval == nullptr' for try_SPILL_COST()

kunalspathak committed:
fix build error

kunalspathak committed:
Call consecutiveCandidates() only for first refposition

kunalspathak committed:
Only perform special handling for non-uppervectorrestore

kunalspathak committed:
jit format

kunalspathak committed:
Add impVectorTableLookup/impVectorTableLookupExtension

kunalspathak committed:
Add the missing break

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
Update assert

kunalspathak committed:
Move definitions in GenTree, fix assert

kunalspathak committed:
fix arm issue

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//  Returns:
+//      Register mask of all the free registers
+//
+regMaskTP LinearScan::getFreeCandidates(regMaskTP candidates, RefPosition* refPosition)
```
Technically yes, but I am passing this to other methods like `getNextConsecutiveRefPosition()` which is also called by passing  `consecutiveRefPosition` which is not const.

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//
+// Arguments:
+//    candidates   - Register assigned to the first refposition.
+//    refPosition  - Number of registers to check.
```
fixed.

kunalspathak committed:
Remove common functions

kunalspathak committed:
Rename info.needsConsecutiveRegisters to info.compNeedsConsecutiveRegisters

kunalspathak committed:
Use needsConsecutiveRegisters template parameter for all configurations

kunalspathak committed:
Handle case of round-robin in getConsecutiveRegisters()

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    assert(firstRefPosition->refType != RefTypeUpperVectorRestore);
+
+    INDEBUG(int refPosCount = 1);
+    regMaskTP busyConsecutiveRegMask = ~(((1ULL << firstRefPosition->regCount) - 1) << firstRegAssigned);
```
I have handled "wrapping" code separately now.

>  In that case, it shouldn't have a ~.

That's right. Removed it.

kunalspathak commented:
> The last pending item is to analyze the TP regression on windows/linux x64:

I do see that `select` instructions increased, but `allocateReg` instructions decreased although all the code is in TARGET_ARM64. There is just one change in `isSpillCandidate` where I moved the fetching of `recentRefPosition` before few checks for better code readability. ARM64 has TP regression that we will have to take given that we need to add checks to decide if `RefPosition` needs consecutive registers or not.


```
??$select@$0A@@RegisterSelection@LinearScan@@QEAA_KPEAVInterval@@PEAVRefPosition@@@Z       : 66555  : NA       : 45.87% : +4.5863%
??$allocateReg@$0A@@LinearScan@@AEAA?AW4_regNumber_enum@@PEAVInterval@@PEAVRefPosition@@@Z : 6242   : NA       : 4.30%  : +0.4301%
??$varTypeUsesFloatReg@W4var_types@@@@YA_NW4var_types@@@Z                                  : -1301  : -100.00% : 0.90%  : -0.0897%
?getRangeEndRef@RefPosition@@QEAAPEAV1@XZ                                                  : -1425  : -100.00% : 0.98%  : -0.0982%
?getRangeEndLocation@RefPosition@@QEAAIXZ                                                  : -1959  : -100.00% : 1.35%  : -0.1350%
?allocateReg@LinearScan@@AEAA?AW4_regNumber_enum@@PEAVInterval@@PEAVRefPosition@@@Z        : -67589 : -100.00% : 46.58% : -4.6576%
```

kunalspathak commented:
@vargaz @fanyang-mono - What is the best way to address the [test failures](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80297-merge-faa8de0a78bb408380/HardwareIntrinsics_ro/1/console.97fd225a.log?helixlogtype=result)? Disable the tests for now, or is it easier to add minimal support?

vargaz commented:
I'd suggest disabling the tests.

kunalspathak committed:
Disable tests for Mono

vargaz commented on src/tests/JIT/HardwareIntrinsics/Arm/Shared/VectorLookupExtension_2Test.template:
```diff
@@ -0,0 +1,400 @@
+                this.inHandle2 = GCHandle.Alloc(this.inArray2, GCHandleType.Pinned);
+                this.inHandle3 = GCHandle.Alloc(this.inArray3, GCHandleType.Pinned);
+                this.outHandle = GCHandle.Alloc(this.outArray, GCHandleType.Pinned);
+
```
This accesses this.outArray which is not set, so its null.

kunalspathak replied:
Good catch. Fixed. I will check why CI didn't complain about it.

kunalspathak replied:
The CI was passing because Arm64 test cases were not running. https://github.com/dotnet/runtime/pull/83948 will fix that.

vargaz commented on src/libraries/System.Runtime.Intrinsics/ref/System.Runtime.Intrinsics.cs:
```diff
@@ -2856,8 +2856,20 @@ public abstract partial class AdvSimd : System.Runtime.Intrinsics.Arm.ArmBase
         public static System.Runtime.Intrinsics.Vector128<ulong> SubtractWideningUpper(System.Runtime.Intrinsics.Vector128<ulong> left, System.Runtime.Intrinsics.Vector128<uint> right) { throw null; }
         public static System.Runtime.Intrinsics.Vector64<byte> VectorTableLookup(System.Runtime.Intrinsics.Vector128<byte> table, System.Runtime.Intrinsics.Vector64<byte> byteIndexes) { throw null; }
         public static System.Runtime.Intrinsics.Vector64<sbyte> VectorTableLookup(System.Runtime.Intrinsics.Vector128<sbyte> table, System.Runtime.Intrinsics.Vector64<sbyte> byteIndexes) { throw null; }
+        public static System.Runtime.Intrinsics.Vector64<byte> VectorTableLookup((Vector128<byte>, Vector128<byte>) table, Vector64<byte> byteIndexes) { throw null; }
```
The implementation uses Vector128<byte>, is this ok ?

kunalspathak replied:
do you mean the `(Vector128<byte>, Vector128<byte>) table` part? It is in accordance with approved APIs in https://github.com/dotnet/runtime/issues/81599.

kunalspathak committed:
Initialize outArray in test

tannergooding commented:
> I'd suggest disabling the tests.

Worth noting that we can't simply disable the tests if Mono is reporting `AdvSimd.IsSupported == true`. This would violate the contract required that all APIs under `AdvSimd` be supported by the underlying runtime.

We'd have to disable the tests and also update `Mono` to report `AdvSimd.IsSupported == false`

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+}
+
+//------------------------------------------------------------------------
+// setNextConsecutiveRegisterAssignment: For subsequent refPositions, set the register
```
`setNext...` seems to imply it only assigns one. Would it be better named `AssignConsecutiveRegisters`?

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    int          registersCount  = firstRefPosition->regCount;
+    RefPosition* nextRefPosition = firstRefPosition;
+    regNumber    regToAssign     = firstRegAssigned;
+    assert(compiler->info.compNeedsConsecutiveRegisters && registersCount > 1);
```
nit: split these into 2 separate asserts?

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//   registers are available in it, and if yes, returns first bit set of every possible series.
+//
+// Arguments:
+//    candidates                - Set of availble candidates.
```
typo
```suggestion
//    candidates                - Set of available candidates.
```

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    {
+        // From LSB, find the first available register (bit `1`)
+        BitScanForward64(&regAvailableStartIndex, static_cast<DWORD64>(currAvailableRegs));
+        regMaskTP startMask = (1ULL << regAvailableStartIndex) - 1;
```
Can `regAvailableStartIndex` ever be `REG_V31` (aka, 63)? Is this expression well-defined in that case?

kunalspathak replied:
Yes and it gets handled by `if (maskProcessed == 0)` case.

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+        {
+            BitScanForward64(&regAvailableEndIndex, static_cast<DWORD64>(maskProcessed));
+        }
+        regMaskTP endMask = (1ULL << regAvailableEndIndex) - 1;
```
Same question here

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+        }
+        regMaskTP endMask = (1ULL << regAvailableEndIndex) - 1;
+
+        // Anything between regAvailableStart and regAvailableEnd is the range of consecutive registers available
```
nit
```suggestion
        // Anything between regAvailableStart and regAvailableEnd is the range of consecutive registers available.
```

BruceForstall commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+        currAvailableRegs &= ~endMask;
+    } while (currAvailableRegs != RBM_NONE);
+
+    if ((candidates & 0x8000000100000000) == 0x8000000100000000)
```
Instead of these hard-coded values, couldn't you use RBM values, e.g., `RBM_V31 | RBM_V0`?

kunalspathak replied:
That's a great suggestion. I will fix it.

BruceForstall reviewed (commented):
some comments

kunalspathak committed:
Add IsSupported checks for VectorLookup/VectorLookupExtension

kunalspathak committed:
Fix the test cases for RunReflectionScenario_UnsafeRead()

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//
+// Arguments:
+//    firstRefPosition  - First refPosition of the series of consecutive registers.
+//    firstRegAssigned  - Register assigned to the first refposition.
```
Not really. We call this method from places where `firstRegAssigned` is the one that corresponds to the Interval of the variable and we will just use it rather than assigning new register to the `firstRefPosition`. In such case `firstRefPosition->registerAssignment` might not reflect the `firstRegAssigned`.

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//
+// Arguments:
+//    firstRefPosition  - First refPosition of the series of consecutive registers.
+//    firstRegAssigned  - Register assigned to the first refposition.
```
Changed all to `RefPosition`.

kunalspathak commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1876,7 +1876,84 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             retNode = impAssignMultiRegTypeToVar(op1, sig->retTypeSigClass DEBUGARG(CorInfoCallConvExtension::Managed));
             break;
         }
+        case NI_AdvSimd_VectorTableLookup:
+        case NI_AdvSimd_Arm64_VectorTableLookup:
```
I agree, it is hard to get a sense of the actual API from the intrinsic names.

kunalspathak commented on src/coreclr/jit/lsra.h:
```diff
@@ -772,8 +775,9 @@ class LinearScan : public LinearScanInterface
-    static const regMaskTP LsraLimitSmallFPSet  = (RBM_V0 | RBM_V1 | RBM_V2 | RBM_V8 | RBM_V9);
+    static const regMaskTP LsraLimitSmallIntSet         = (RBM_R0 | RBM_R1 | RBM_R2 | RBM_R19 | RBM_R20);
+    static const regMaskTP LsraLimitSmallFPSet          = (RBM_V0 | RBM_V1 | RBM_V2 | RBM_V8 | RBM_V9);
+    static const regMaskTP LsraExtraFPSetForConsecutive = (RBM_V3 | RBM_V5 | RBM_V7);
```
Renamed to `LsraLimitFPSetForConsecutive`.

kunalspathak committed:
Review feedback

kunalspathak committed:
wip

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
Merge branch 'tbl_tbx_temp' into tbl_tbx

kunalspathak commented:
I realized that in my recent refactoring, I removed the heuristics that picks the series based on least registers to spill. I will add it back.

kunalspathak committed:
fix a typo in test case

kunalspathak committed:
Add filterConsecutiveCandidatesForSpill() to select range that needs fewer register spilling

kunalspathak commented:
> I realized that in my recent refactoring, I removed the heuristics that picks the series based on least registers to spill. I will add it back.

Done.

kunalspathak commented:
/azp runtime-coreclr jitstressregs

kunalspathak commented:
/azp run runtime-coreclr jitstressregs

vargaz committed:
Add mono support.

vargaz commented:
This should fix/improve mono support:
https://github.com/vargaz/runtime/commit/f247b3c6408c60b751999ca30710005d2db9a044

kunalspathak commented:
> This should fix/improve mono support: [vargaz@f247b3c](https://github.com/vargaz/runtime/commit/f247b3c6408c60b751999ca30710005d2db9a044)

Thanks a lot @vargaz for quick fix. Appreciate it.

kunalspathak commented:
> /azp run runtime-coreclr jitstressregs

Looks like all the `VectorTableLookupExtension` cases are failing because they are operating on default values. I will investigate and submit a fix.

kunalspathak commented:
> > /azp run runtime-coreclr jitstressregs
> 
> Looks like all the `VectorTableLookupExtension` cases are failing because they are operating on default values. I will investigate and submit a fix.

It seems there is a extra `mov` added from `defaultValues` to `row1` which is why we are seeing the mismatch:

```
 defaultValues: (113, 93, 212, 27, 58, 37, 6, 118, 139, 118, 173, 30, 18, 53, 165, 236)
 firstOp: (100, 158, 153, 151, 135, 230, 102, 106, 199, 113, 65, 180, 152, 242, 233, 25)
 secondOp: (191, 98, 103, 254, 0, 35, 2, 121, 98, 115, 15, 220, 239, 184, 9, 46)
 indices: (33, 8, 7, 9, 0, 8, 34, 36, 2, 25, 10, 28, 19, 37, 4, 36)
 result: (113, 139, 118, 118, 113, 139, 6, 118, 212, 115, 173, 239, 254, 53, 58, 236)
```

<img width="1578" alt="image" src="https://user-images.githubusercontent.com/12488060/229050033-60c24c59-52a5-435a-8dfd-599a473588e6.png">

Will investigate more.

kunalspathak committed:
Delay free the registers for VectorTableLookupExtension

kunalspathak committed:
Merge commit 'f247b3c6408c60b751999ca30710005d2db9a044' into tbl_tbx

kunalspathak commented:
> It seems there is a extra `mov` added from `defaultValues` to `row1` which is why we are seeing the mismatch:

This was because we were overwriting the targetReg into the operand of the operation. I had delayRegFree logic for VectorTableLookupExtension but thought was not necessary at one point. Added it back.

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into tbl_tbx

kunalspathak committed:
fix mono build error

kunalspathak commented:
@BruceForstall - this should be ready to review again.

BruceForstall reviewed (approved):
LGTM

kunalspathak commented:
failures are known issues. Thanks @BruceForstall for the review.
